integrations: Add webhook code, API endpoint, tests, and documentation for Facebook.

nyan-salmon opened the pull request:
Facebook webhook has many webhook items.
So I implemented webhooks included in the fixtures folder.
(https://chat.zulip.org/#narrow/stream/GCI.20help/subject/Write.20Facebook.20integration/near/422421)

---

And I think [this commit](https://github.com/zulip/zulip/pull/7975/commits/35f933a25c901198ebce90e58a249e288076e22a) is wrong.
So, I added correct fixtures.
(https://chat.zulip.org/#narrow/stream/GCI.20help/subject/Write.20Facebook.20integration/near/430812)

---

Receiving Facebook webhook requires HTTPS connection.
If you want test my pull request on test environment which provided by zulip community, Please use [my script](https://github.com/nyan-salmon/fb-webhooks-forward).
(https://chat.zulip.org/#narrow/stream/GCI.20help/subject/Write.20Facebook.20integration/near/430043)

---

zerver/webhooks/facebook/view.py Line 83~91
```python
    tmp_body = str(payload)
    tmp_body = tmp_body.replace("\'", '\"')
    tmp_body = tmp_body.replace('["', '!#!$!%!&!')
    tmp_body = tmp_body.replace('"]', '!&!%!$!#!')
    tmp_body = tmp_body.replace('[', '')
    tmp_body = tmp_body.replace(']', '')
    tmp_body = tmp_body.replace("!#!$!%!&!", '["')
    tmp_body = tmp_body.replace("!&!%!$!#!", '"]')
    _payload = json.loads(tmp_body)
```
I think that this code is not good, but a reply from Facebook contains an insignificant array, and without this code, the following code can not be executed.
```python
payload['entry']['changes']['value']
```
Please tell me if you have other good ways.

---

Thank you for reading!

eeshangarg commented on zerver/webhooks/facebook/fixtures/permissions_manage_pages.json:
```diff
@@ -1,13 +1,22 @@
 {
-   "entry":[
```
@nyan-salmon: Could you please explain in detail why you changed all of these fixtures? :)

nyan-salmon replied:
Sorry, I was not good.

I fixed it.

I changed some files name to conform to the facebook naming convention.

eeshangarg commented on zerver/webhooks/facebook/view.py:
```diff
@@ -0,0 +1,123 @@
+
+def permissions_object(payload: Dict[Text, Any]) -> Text:
+    field = payload['entry']['changes']['field']
+    message = '**{_field} permission** changed'.format(_field=field)
```
I would suggest just using `field=field` instead of `_field`.

nyan-salmon replied:
I didn't know that can use same name.

I fixed it!

eeshangarg commented on zerver/webhooks/facebook/view.py:
```diff
@@ -0,0 +1,123 @@
+def api_facebook_webhook(request: HttpRequest, user_profile: UserProfile,
+                        stream: Text=REQ(default='Facebook'), token: Text=REQ(),
+                        payload: Dict[Text, Any]=REQ(argument_type='body')) -> HttpResponse:
+    tmp_body = str(payload)
```
I don't think you need to convert this to a `str`.

nyan-salmon replied:
If I remove this converting, script make error as cannot edit Dict Object.

Is there any other way?

eeshangarg commented on zerver/webhooks/facebook/view.py:
```diff
@@ -0,0 +1,123 @@
+    tmp_body = tmp_body.replace(']', '')
+    tmp_body = tmp_body.replace("!#!$!%!&!", '["')
+    tmp_body = tmp_body.replace("!&!%!$!#!", '"]')
+    _payload = json.loads(tmp_body)
```
Again, I am not convinced that you need all of this, your payloads look fine to me and shouldn't need this code to sanitize the JSON. Could you explain further as to why you are doing this?

nyan-salmon replied:
Removing this method like [view.py on my Gist](https://gist.github.com/nyan-salmon/271c7a67ec606ec33798f1de8f5d7905) gives the following error.

Cannot purse as `payload['entry']['changes']['value']['message']` because it is inserted in an unnecessary array.

```
(zulip-py3-venv) ~/zulip2 (master)$ ./tools/test-backend zerver/webhooks/facebook
-- Running tests in serial mode.
Destroying test database for alias 'default'...
Using existing clone for alias 'default'...
Running zerver.webhooks.facebook.tests.FacebookTests.test_application_plugin_comment
2018-01-11 15:22:08.949 ERR  [] Traceback (most recent call last):
  File "/srv/zulip-py3-venv/lib/python3.5/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/srv/zulip-py3-venv/lib/python3.5/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/home/zulipdev/zulip2/zerver/lib/request.py", line 170, in _wrapped_view_func
    return view_func(request, *args, **kwargs)
  File "/home/zulipdev/zulip2/zerver/decorator.py", line 306, in _wrapped_func_arguments
    raise err
  File "/home/zulipdev/zulip2/zerver/decorator.py", line 280, in _wrapped_func_arguments
    return view_func(request, user_profile, *args, **kwargs)
  File "/home/zulipdev/zulip2/zerver/lib/request.py", line 170, in _wrapped_view_func
    return view_func(request, *args, **kwargs)
  File "/home/zulipdev/zulip2/zerver/webhooks/facebook/view.py", line 104, in api_facebook_webhook
    body = get_body_based_on_object(object)(_payload)
  File "/home/zulipdev/zulip2/zerver/webhooks/facebook/view.py", line 58, in application_object
    field = payload['entry']['changes']['field']
TypeError: list indices must be integers or slices, not str


======================================================================
FAIL: test_application_plugin_comment (zerver.webhooks.facebook.tests.FacebookTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/zulipdev/zulip2/zerver/webhooks/facebook/tests.py", line 18, in test_application_plugin_comment
    expected_subject, expected_message)
  File "/home/zulipdev/zulip2/zerver/lib/test_classes.py", line 633, in send_and_test_stream_message
    self.STREAM_NAME, **kwargs)
  File "/home/zulipdev/zulip2/zerver/lib/test_classes.py", line 573, in send_json_payload
    self.assert_json_success(result)
  File "/home/zulipdev/zulip2/zerver/lib/test_classes.py", line 449, in assert_json_success
    self.assertEqual(result.status_code, 200, json['msg'])
AssertionError: 500 != 200 : Internal server error

----------------------------------------------------------------------
Ran 1 test in 0.253s

FAILED (failures=1)
Destroying test database for alias 'default'...
```

nyan-salmon replied:
I wrote that you had to delete unnecessary arrays earlier, however for example, the target_ids array included in `fixtures/permissions_ads_management.json` is necessary.

So, the code as shown below doesn't work parfectly.
```
tmp_body = str(payload)
tmp_body = tmp_body.replace('[', '')
tmp_body = tmp_body.replace(']', '')
```

To protect a necessary array, We need the code as I commited.

eeshangarg replied:
Okay, there are a couple of points I'd like you to consider:

* Converting a payload dictionary to a `str` and then loading it again with `json.loads` is not right, so remove all of that code and the `tmp_body.replace` lines.
* Once you remove all of that code and run into that error, try to fix the error by not changing the `payload` but by figuring out how you can get to the data you want. So the error you showed me seems to arise when you try to give a key where it expects a list index, i.e., a number, so look at the payload and see how you can get to the value you want using an index without removing the "unnecessary array".

A good thing to keep in mind is that you should never modify a payload's content, you should just access it and use the data in whatever way you want. :)

nyan-salmon commented:
And I think that `test-all` does not succeed because there is no `doc.md` file.

Thank you for reading and reviewing!

eeshangarg commented on zerver/webhooks/facebook/view.py:
```diff
@@ -0,0 +1,123 @@
+def api_facebook_webhook(request: HttpRequest, user_profile: UserProfile,
+                         stream: Text=REQ(default='Facebook'), token: Text=REQ(),
+                         payload: Dict[Text, Any]=REQ(argument_type='body')) -> HttpResponse:
+    tmp_body = str(payload)
```
Okay, first of all, you don't need to convert `payload` to a `str`. `payload` is a dictionary and I would recommend leaving it as it is and dealing with it directly.

eeshangarg replied:
Also, if you don't convert it to a `str`, you won't have to call `_payload = json.loads(tmp_body)` either. So remove that too.

nyan-salmon commented:
@eeshangarg Thank you!
Would you please review again?

nyan-salmon commented:
@eeshangarg Thank you!
I checked it before sending a pull request because an error which does not occur with the local test-all command occurs in travis-ci.

I fixed some errors, but I can not solve the currently occurring error.
https://travis-ci.org/nyan-salmon/zulip/jobs/328427751

What is the cause?
And I'm sorry many times...

alenavolk commented:
@nyan-salmon I think this is a bug that was fixed in the upstream repository a couple of days ago: b403897. Updating your fork should fix the problem :)

nyan-salmon commented:
@alenavolk Thank you! and travis-ci problems were fixed!
But circleci cause an another error(frontend)...
Maybe it seems not my change that caused the error, Is this also a bug?

eeshangarg commented on zerver/webhooks/facebook/doc.md:
```diff
@@ -0,0 +1,3 @@
+{!create-stream.md!}
+
+{!create-bot-construct-url.md!}
```
Now, I would recommend studying the documentation for other integrations and then writing the contents of the `doc.md` file. And claim the documentation task for the Facebook integration. We can't merge the PR till it is documented! :)

eeshangarg replied:
And don't worry about circleci, it is not because of your PR! :)

alenavolk replied:
@nyan-salmon This is the task for adding documentation: [5647007018385408](https://codein.withgoogle.com/tasks/5647007018385408), I think it would be perfect for you. But be sure to claim it soon, the claim new task deadline is just around the corner :)

nyan-salmon replied:
@eeshangarg @alenavolk Thank you for advices!
I will claim that task!

nyan-salmon commented:
Added `doc.md` and changed PR title/commit message.

I’m not good at English, so please tell me if I have the wrong grammar!

Thank you very much for waiting for a long time.

sampritipanda commented:
@nyan-salmon Does the logo show up on the integrations page on your development server?

It doesn't work on my server, take a look here: http://sampritipanda.zulipdev.org:9991/integrations/doc/facebook

or maybe it's something to do with svg not working on my server, if that is the case I think it would be better to use a .png file instead of .svg.

sampritipanda commented on zerver/webhooks/facebook/doc.md:
```diff
@@ -0,0 +1,82 @@
+Then, fill that form to create a new Facebook app.
+
+In `Display Name` column, enter the name as you can distinguishable your app.  
+And `Contact Email` column, enter the name as you can receive e-mail from Facebook.
```
You can delete these two lines (40, 41), the screenshot is enough.

eeshangarg replied:
Yep agreed! Delete these two lines, you don't need them! :)

sampritipanda commented on zerver/webhooks/facebook/doc.md:
```diff
@@ -0,0 +1,82 @@
+
+![](/static/images/integrations/facebook/003.png)
+
+Now, you can see the page like the image.  
```
This line can be removed too.

eeshangarg replied:
You don't need this line, I would suggest removing it! :)

eeshangarg commented on zerver/webhooks/facebook/doc.md:
```diff
@@ -0,0 +1,82 @@
+
+<!--
+{!create-bot-construct-url.md!}
+-->
```
Could you delete this comment please?

eeshangarg commented on zerver/webhooks/facebook/doc.md:
```diff
@@ -0,0 +1,82 @@
+of your Zulip bot, and `stream` is the stream name you want the
+notifications sent to, and `token` will be used for server verification later,
+please decide your favorite character string.
+
```
Make `token will be used for server....` a separate sentence. :)

eeshangarg replied:
Also, make it `please choose your favorite character string.` And could you please explain as to what you mean by this? Maybe elaborate a little more?

nyan-salmon replied:
@eeshangarg 
I tried to think of it as a friend at school, but I did not know how to fix it.

Would you please tell me how to fix it?

eeshangarg commented on zerver/webhooks/facebook/doc.md:
```diff
@@ -0,0 +1,82 @@
+
+### Configuring the webhook
+
+Go and Sign In to the following URL: <https://developers.facebook.com/apps/>
```
Change this to `Sign in at the following URL:`

eeshangarg commented on zerver/webhooks/facebook/doc.md:
```diff
@@ -0,0 +1,82 @@
+
+![](/static/images/integrations/facebook/001.png)
+
+Then, fill that form to create a new Facebook app.
```
Change this to `Then, fill the following form to create a new Facebook app:`

eeshangarg commented on zerver/webhooks/facebook/doc.md:
```diff
@@ -0,0 +1,82 @@
+
+![](/static/images/integrations/facebook/002.png)
+
+Next, put your mouse cursor over **Webhooks** item, and click on **Set up** button.
```
Change this to `Next, under **Webhooks**, click on **Set up**:`

eeshangarg commented on zerver/webhooks/facebook/doc.md:
```diff
@@ -0,0 +1,82 @@
+![](/static/images/integrations/facebook/003.png)
+
+Now, you can see the page like the image.  
+To choose a category for webhook, click on the drop down list.
```
Change this to `Choose a category for the webhook:`. That's all you need, since the screenshots pretty much make it clear what's happening :)

eeshangarg commented on zerver/webhooks/facebook/doc.md:
```diff
@@ -0,0 +1,82 @@
+
+![](/static/images/integrations/facebook/004.png)
+
+This guide explains the procedure of subscribe to "feed" in the "user" category.
```
Change this to `This guide explains how to subscribe to a "feed" in the **User** category.`

eeshangarg commented on zerver/webhooks/facebook/doc.md:
```diff
@@ -0,0 +1,82 @@
+
+This guide explains the procedure of subscribe to "feed" in the "user" category.
+
+Chose **User** from the drop down list, and click on **Subscribe to this topic** button.  
```
Change this to `Select the **User** category, and click on **Subscribe to this topic**`

eeshangarg commented on zerver/webhooks/facebook/doc.md:
```diff
@@ -0,0 +1,82 @@
+And then, click on **Verify and Save** button.
+
+![](/static/images/integrations/facebook/005.png)
+
```
Okay, so all of the above should be replaced by this (Replace everything from line 57 to 67 exactly by what I type up below):
```
Fill the **Edit User Subscription** form as follows:

1. **Callback URL**: enter the webhook URL created above.
2. **Verify Token**: enter the token you chose above. For instance, in this example you may enter **sampletoken**)
3. Activate the **Include Values** option.
4. Click on **Verify and Save**.

The resulting form would look like:

![](/static/images/integrations/facebook/005.png)
```

eeshangarg commented on zerver/webhooks/facebook/doc.md:
```diff
@@ -0,0 +1,82 @@
+![](/static/images/integrations/facebook/005.png)
+
+Finally, click on **Subscribe** button and **Test** button in the **feed** line.
+
```
Change this to `Finally, click **Subscribe** and **Test** in the **feed** row, like so:`

eeshangarg commented on zerver/webhooks/facebook/doc.md:
```diff
@@ -0,0 +1,82 @@
+
+And click on **Send to My Server** button.  
+Now the test message will be sent to your Zulip server.
+
```
Replace these two lines by:
```
Click on **Send to My Server** and a test message will be sent to your Zulip server.
```

eeshangarg commented:
@nyan-salmon: Also, don't worry about an approval, I know the deadline is getting closer. I'll make sure I approve this, but to have the PR merged, you'll have to implement my feedback! Thanks! :)

nyan-salmon commented:
@sampritipandaThank you for reporting logo problem.

I tested in Google Chrome(Win/Mac/Ubuntu/iOS/Android), Safari(Mac/iOS), Edge and Edge.

As a result, Chrome, Safari and Edge successfully shown the logo, but it did not show properly in IE.
But that result was the same for other integration using SVG for the logo.

Is it better to replace the logo with PNG?

The IE display can be viewed from the following links.
https://i.imgur.com/oWXk1tl.png

nyan-salmon commented:
@sampritipanda @eeshangarg
Thank you very much for reviewing:tada:

I applied almost advices!

sampritipanda commented:
@nyan-salmon Yes, I think it would be best to use a .png logo. I was using Firefox and the svg logo didn't work. We can replace the svg logos for other integrations in a different PR.

sampritipanda commented on zerver/webhooks/facebook/doc.md:
```diff
@@ -0,0 +1,74 @@
+
+![](/static/images/integrations/facebook/001.png)
+
+Then, fill that the following form to create a new Facebook app:
```
"fill in the"

sampritipanda commented on zerver/webhooks/facebook/doc.md:
```diff
@@ -0,0 +1,74 @@
+Fill the **Edit User Subscription** form as follows:
+
+1. **Callback URL**: enter the webhook URL created above.
+2. **Verify Token**: enter the token you chose above. For instance, in this example you may enter **sampletoken**)
```
There is a extra bracket `)` at the end,

eeshangarg commented:
@nyan-salmon: I think this looks good to me for now, so I'll approve this on GCI! I'll give this a one final review before we merge though! Thanks for working on this! :)

nyan-salmon commented:
@sampritipanda Thank you for reviewing!
I fixed them!

eeshangarg commented on zerver/webhooks/facebook/view.py:
```diff
@@ -0,0 +1,114 @@
+            if request.GET.get('hub.verify_token') == token:
+                return HttpResponse(request.GET.get('hub.challenge'))
+            else:
+                return json_error(('Error: Token is wrong'))  # nocoverage
```
Two things:
* We should probably write a test for this and provide a fixture and not use `#  nocoverage` unless there is a good reason that we can't test this :)

eeshangarg commented on zerver/webhooks/facebook/view.py:
```diff
@@ -0,0 +1,114 @@
+            else:
+                return json_error(('Error: Token is wrong'))  # nocoverage
+        else:
+            return json_error(('Error: Unsupported method'))  # nocoverage
```
Same as above, I would test this! :)

eeshangarg commented on zerver/webhooks/facebook/view.py:
```diff
@@ -0,0 +1,114 @@
+            subject = object + " notification"
+            check_send_stream_message(user_profile, request.client,
+                                      stream, subject, body)
+        return json_success()
```
I would use `event` instead of `object` and `get_event` instead of `get_object` for two reasons:
* `object` is a Python keyword, so we should probably never use it.
* `event` and `get_event` are more consistent with how similar functions are named in our other webhook implementations.

eeshangarg commented on zerver/webhooks/facebook/view.py:
```diff
@@ -0,0 +1,114 @@
+                                      stream, subject, body)
+        return json_success()
+
+OBJECTS_FUNCTION_MAPPER = {
```
Again, you can name this `EVENTS_FUNCTION_MAPPER` because that's what we use all over our codebase. This is not a huge deal though! :)

eeshangarg commented on zerver/webhooks/facebook/view.py:
```diff
@@ -0,0 +1,114 @@
+    object = payload['object']
+    if object in OBJECTS_FUNCTION_MAPPER:
+        return object
+    raise UnknownObjectType(u"Object '{}' is unknown and cannot be handled".format(object))  # nocoverage
```
Same advice as above:
1. Use `event` instead of `object` and `get_event` instead of `get_object`.
2. Any particular reason, why we aren't testing for that last line?

nyan-salmon replied:
2.
Maybe, this case return 500 error.
Is there a function to check 500 errors?

This integration was written with reference to [zerver/webhooks/gci](https://github.com/zulip/zulip/tree/master/zerver/webhooks/gci), but it looks like use # nocoverage...
Do I also have to fix this?

eeshangarg replied:
@nyan-salmon: No worries, don't worry about it! I just ran a quick `git grep` and it turns out that last line isn't tested anywhere anyway, so it is not a huge deal! :)

eeshangarg commented on zerver/webhooks/facebook/view.py:
```diff
@@ -0,0 +1,114 @@
+    raise UnknownObjectType(u"Object '{}' is unknown and cannot be handled".format(object))  # nocoverage
+
+def get_body_based_on_object(object: Text) -> Any:
+    return OBJECTS_FUNCTION_MAPPER[object]
```
Rename as `get_body_based_on_event`. :)

nyan-salmon commented:
@eeshangarg: Thank you for reviewing 😄 

Updated!

timabbott commented on zerver/webhooks/facebook/doc.md:
```diff
@@ -0,0 +1,74 @@
+
+Modify the parameters of the URL above, where `api_key` is the API key
+of your Zulip bot, and `stream` is the stream name you want the
+notifications sent to.  
```
Why isn't this using the macro `create-bot-construct-url.md`?  It looks like the only different is the URL; it seems like we could address that without the duplication by just adding a new variable to `webhook-url.md` so it looks like this: 
`{{ api_url }}{{ integration_url }}?api_key=abcdefgh&stream={{ recommended_stream_name }}{{ custom_params }}`    .

Also, I don't understand the `token` -- should it be a secure, randomly generated string, a label for what the integration is that will appear in the Facebook UI, or something else?  "favorite character" string doesn't seem like clear advice for this sort of professional context.

nyan-salmon replied:
Sorry, I can't understand markdown macro even if I read this page(https://chat.zulip.org/api/integration-docs-guide#markdown-macros).

Would you please tell me a bit more specific procedure if possible?

eeshangarg replied:
@timabbott: Agreed about the macro, I noticed that too but the only reason I didn't suggest that was because adding support for custom params in the webhook URL macro is out of the scope of this PR and should probably be a separate PR altogether. It would most likely involve setting integration-specific context values in the `Integration` instance or something. Not to mention, such a change would apply to all integrations, not just this one! So, it seems like a project of its own!

@nyan-salmon: I agree that the `token` string should be explained in more detail. Could you please investigate further and see if Facebook's docs have anything on that?

timabbott replied:
@eeshangarg I think it's reasonable to leave this for a follow-up issue.

timabbott commented on zerver/webhooks/facebook/view.py:
```diff
@@ -0,0 +1,114 @@
+            if request.GET.get('hub.verify_token') == token:
+                return HttpResponse(request.GET.get('hub.challenge'))
+            else:
+                return json_error(('Error: Token is wrong'))
```
What's with the extra parens around the `()` argument to `json_error` here?  We should be passing a string tagged for translation using `_`.  Can you explain why you did this?  I'm wondering if the linter errors need more clarity.

timabbott commented on zerver/webhooks/facebook/view.py:
```diff
@@ -0,0 +1,114 @@
+                return json_error(('Error: Token is wrong'))
+        else:
+            return json_error(('Error: Unsupported method'))
+    else:  # facebook webhook receiver
```
It looks like this `else` statement isn't required because all of the code paths above return; can you clean that up?  I think there's some similar unnecessary `else return` blocks above, too.

timabbott commented on zerver/webhooks/facebook/view.py:
```diff
@@ -0,0 +1,114 @@
+        return json_success()
+
+EVENTS_FUNCTION_MAPPER = {
+    'user': user_event,
```
It looks like there may be more event types that we don't support; can you add a comment just oting that this is the case?

timabbott commented:
@eeshangarg I posted a few quick comments on things still needing work, otherwise lgtm.

nyan-salmon commented:
@eeshangarg: I'm sorry for asking your advice so many times.

Would you please review it again?

nyan-salmon commented:
@eeshangarg: updated doc.md

eeshangarg commented on zerver/webhooks/facebook/doc.md:
```diff
@@ -0,0 +1,73 @@
+of your Zulip bot, and `stream` is the stream name you want the
+notifications sent to.  
+
+`token` will be used for server verification later, please generate randomly string.
```
@nyan-salmon: Okay so I checked Facebook's API docs and this should be replaced with:
```
`token` is an arbitrary string of your choosing that can be used to confirm to your
 server that the request is valid. This string will be included in Facebook's
 incoming payloads each time they send your server a verification request.
```

nyan-salmon replied:
@eeshangarg: 
Sorry...
I have forgot run "git rebase --continue", so My modify was not reflected.

I changed your description from mine.

nyan-salmon replied:
My explanation was as follow:

```
When you enable Webhook in Facebook console, this integration receive a verification request from Facebook.
And this integration will verify `token` matches the value you will enter on Facebook console later.

This is a security check to authenticate the request and identify the Webhook.
So, please generate randomly string.
```

timabbott commented:
Merged after removing a couple bits of trailing whitespace, thanks @nyan-salmon and @eeshangarg!  I imagine we'll end up reworking the docs significantly as part of the larger integration-docs rewrite @eeshangarg is working on to add numbered lists, but that's fine to do as a follow-up since the migration is still in progress. :)
